core/src/frame_lifecycle.rs

```diff
@@ -73,63 +72,35 @@ pub fn run_all_phases_avm2(context: &mut UpdateContext<'_, '_>) {
     let stage = context.stage;
 
     *context.frame_phase = FramePhase::Enter;
-
-    stage.enter_frame(context);
-
-    let mut ran_framescript = FnvHashSet::default();
-
     Avm2::each_orphan_movie(context, |movie, context| {
-        if movie.initialized() {
-            // Orphan frame scripts run in a really weird place.
-            // Running them here matches the output we get from Flash Player,
-            // where the currentFrame field field hasn't been updated yet
-            // (but is updated when we call an enterFrame listener)
-            movie.run_frame_scripts(context);
-            movie.enter_frame(context);
-            ran_framescript.insert(movie.downgrade().as_ptr());
-        }
+        movie.enter_frame(context);
     });
+    stage.enter_frame(context);
 
     *context.frame_phase = FramePhase::Construct;
-
-    Avm2::each_orphan_movie(context, |movie, context| {
-        if !movie.initialized() {
-            movie.run_frame_scripts(context);
-            ran_framescript.insert(movie.downgrade().as_ptr());
-        }
-    });
-
-    stage.construct_frame(context);
-
     Avm2::each_orphan_movie(context, |movie, context| {
         movie.construct_frame(context);
-        if !ran_framescript.contains(&movie.downgrade().as_ptr()) {
-            movie.run_frame_scripts(context);
-        }
     });
-
+    stage.construct_frame(context);
     stage.frame_constructed(context);
 
     *context.frame_phase = FramePhase::FrameScripts;
-
-    stage.run_frame_scripts(context);
-
     Avm2::each_orphan_movie(context, |movie, context| {
-        if !movie.initialized() {
-            movie.enter_frame(context);
-        }
+        movie.run_frame_scripts(context);
     });
+    stage.run_frame_scripts(context);
 
     *context.frame_phase = FramePhase::Exit;
-    stage.exit_frame(context);
-
     Avm2::each_orphan_movie(context, |movie, context| {
         movie.on_exit_frame(context);
     });
+    stage.exit_frame(context);
```

Comment:
Is the changed order of `exit_frame` compared to orphan movies' `exit_frame` correct? (this also applies to `run_frame_scripts`)

Reply:
I don't think the order is currently observable - all of the tests pass with this code.